File: mobile_app/lib/services/speaker_assembly.dart
// Monta speaker_view dai segmenti Whisper + speaker_ids dell'LLM.

String formatMmSs(num seconds) {
  final total = seconds.isNaN ? 0 : seconds.round().clamp(0, 24 * 3600);
  final minutes = (total ~/ 60).toString().padLeft(2, '0');
  final secs = (total % 60).toString().padLeft(2, '0');
  return '$minutes:$secs';
}

List<int> normalizeSpeakerIds(dynamic value, int segmentCount) {
  if (segmentCount <= 0) return const [];
  if (value is! List) return List<int>.filled(segmentCount, 0);

  final ids = <int>[];
  for (final item in value) {
    final parsed = item is num ? item.toInt() : int.tryParse('$item') ?? 0;
    ids.add(parsed < 0 ? 0 : parsed);
  }

  if (ids.length < segmentCount) {
    final pad = ids.isEmpty ? 0 : ids.last;
    ids.addAll(List<int>.filled(segmentCount - ids.length, pad));
  } else if (ids.length > segmentCount) {
    return ids.sublist(0, segmentCount);
  }
  return ids;
}

List<Map<String, String>> buildSpeakerView(
  List<Map<String, dynamic>> segments,
  dynamic speakerIds,
) {
  if (segments.isEmpty) return const [];

  final ids = normalizeSpeakerIds(speakerIds, segments.length);
  final blocks = <Map<String, String>>[];

  var currentSpeaker = ids[0];
  var currentStart = (segments[0]['start'] as num?)?.toDouble() ?? 0.0;
  final texts = <String>[];

  void flush() {
    final joined = texts.where((t) => t.isNotEmpty).join(' ').trim();
    texts.clear();
    if (joined.isEmpty) return;
    blocks.add({
      'speaker': 'Speaker $currentSpeaker',
      'text': joined,
      'time': formatMmSs(currentStart),
    });
  }

  for (var i = 0; i < segments.length; i++) {
    final text = (segments[i]['text'] as String? ?? '').trim();
    final speaker = ids[i];
    final start = (segments[i]['start'] as num?)?.toDouble() ?? 0.0;

    if (speaker != currentSpeaker && texts.isNotEmpty) {
      flush();
      currentSpeaker = speaker;
      currentStart = start;
    }

    if (texts.isEmpty) {
      currentSpeaker = speaker;
      currentStart = start;
    }
    if (text.isNotEmpty) texts.add(text);
  }
  flush();
  return blocks;
}

String speakerViewToFormatted(List<Map<String, String>> speakerView) {
  if (speakerView.isEmpty) return '';
  return speakerView.map((block) {
    final time = block['time'];
    final timeSuffix = (time != null && time.isNotEmpty) ? ' - $time' : '';
    return '[${block['speaker']}$timeSuffix]: ${block['text']}';
  }).join('\n');
}

({List<Map<String, String>> speakerView, String formatted})
    buildFromRawTranscript(String transcript) {
  final text = transcript.trim();
  if (text.isEmpty) {
    return (speakerView: <Map<String, String>>[], formatted: '');
  }
  final view = [
    {'speaker': 'Speaker 0', 'text': text, 'time': '00:00'},
  ];
  return (speakerView: view, formatted: speakerViewToFormatted(view));
}

String formatSegmentsForDiarization(
  List<Map<String, dynamic>> segments, {
  int maxCharsPerSegment = 280,
}) {
  final lines = <String>[];
  for (var i = 0; i < segments.length; i++) {
    final start = formatMmSs((segments[i]['start'] as num?) ?? 0);
    var text = (segments[i]['text'] as String? ?? '')
        .trim()
        .replaceAll('\n', ' ');
    if (text.length > maxCharsPerSegment) {
      text = '${text.substring(0, maxCharsPerSegment - 1)}…';
    }
    lines.add('[$i] [$start] $text');
  }
  return lines.join('\n');
}

File: backend/services/speaker_assembly.py
"""Monta speaker_view e formatted_transcript dai segmenti Whisper.

L'LLM fa solo diarizzazione compatta (chi parla in ciascun segmento); il testo
resta quello verbatim di Whisper, cosi' non si riassume e non si riscrive
tutta la trascrizione in output.
"""

from __future__ import annotations

from typing import Any


def format_mmss(seconds: float) -> str:
    total = max(0, int(seconds))
    return f"{total // 60:02d}:{total % 60:02d}"


def normalize_speaker_ids(value: Any, segment_count: int) -> list[int]:
    """Allinea speaker_ids alla lunghezza dei segmenti; fallback tutto Speaker 0."""
    if segment_count <= 0:
        return []
    if not isinstance(value, list):
        return [0] * segment_count

    ids: list[int] = []
    for item in value:
        try:
            ids.append(max(0, int(item)))
        except (TypeError, ValueError):
            ids.append(0)

    if len(ids) < segment_count:
        ids.extend([ids[-1] if ids else 0] * (segment_count - len(ids)))
    elif len(ids) > segment_count:
        ids = ids[:segment_count]
    return ids


def build_speaker_view(
    segments: list[dict[str, Any]],
    speaker_ids: list[int] | None = None,
) -> list[dict[str, str]]:
    """Unisce segmenti consecutivi dello stesso speaker in blocchi verbatim."""
    if not segments:
        return []

    ids = normalize_speaker_ids(speaker_ids, len(segments))
    blocks: list[dict[str, str]] = []

    current_speaker = ids[0]
    current_start = float(segments[0].get("start") or 0.0)
    texts: list[str] = []

    def flush() -> None:
        nonlocal texts
        joined = " ".join(t for t in texts if t).strip()
        texts = []
        if not joined:
            return
        blocks.append(
            {
                "speaker": f"Speaker {current_speaker}",
                "text": joined,
                "time": format_mmss(current_start),
            }
        )

    for index, segment in enumerate(segments):
        text = str(segment.get("text", "")).strip()
        speaker = ids[index]
        start = float(segment.get("start") or 0.0)

        if speaker != current_speaker and texts:
            flush()
            current_speaker = speaker
            current_start = start

        if not texts:
            current_speaker = speaker
            current_start = start
        if text:
            texts.append(text)

    flush()
    return blocks


def speaker_view_to_formatted(speaker_view: list[dict[str, str]]) -> str:
    if not speaker_view:
        return ""
    parts: list[str] = []
    for block in speaker_view:
        time_value = block.get("time")
        time_suffix = f" - {time_value}" if time_value else ""
        parts.append(f"[{block['speaker']}{time_suffix}]: {block['text']}")
    return "\n".join(parts)


def build_from_raw_transcript(transcript: str) -> tuple[list[dict[str, str]], str]:
    """Fallback senza segmenti: un solo blocco con tutta la trascrizione grezza."""
    text = transcript.strip()
    if not text:
        return [], ""
    view = [{"speaker": "Speaker 0", "text": text, "time": "00:00"}]
    return view, speaker_view_to_formatted(view)


def format_segments_for_diarization(
    segments: list[dict[str, Any]], *, max_chars_per_segment: int = 280
) -> str:
    """Elenco numerato leggero da passare all'LLM per la sola diarizzazione."""
    lines: list[str] = []
    for index, segment in enumerate(segments):
        start = format_mmss(float(segment.get("start") or 0.0))
        text = str(segment.get("text", "")).strip().replace("\n", " ")
        if len(text) > max_chars_per_segment:
            text = text[: max_chars_per_segment - 1] + "…"
        lines.append(f"[{index}] [{start}] {text}")
    return "\n".join(lines)
